webapps/ui/cockpit/client/scripts/components/callActivityOverlays/definition/index.js

```diff
@@ -31,72 +32,128 @@ module.exports = function(viewContext) {
-                  activity,
-                  calledProcess,
-                  overlays,
-                  clickListener,
-                  toolTipTitle,
-                  $scope,
-                  $timeout
-                );
-              }
-            };
-
-            drawStaticLinks(staticCalledProcessDefinitions);
+            PluginProcessDefinitionResource.getCalledProcessDefinitions({
+              id: processDefinition.id
+            })
+              .$promise.then(dynamicCalledProcessDefinitions =>
+                drawLinks(
+                  staticCalledProcessDefinitions,
+                  dynamicCalledProcessDefinitions
+                )
+              )
+              .catch(angular.noop);
           }
         );
       }
+      const drawLinks = function(staticProcDefs, dynamicProcDefs) {
+        const callActivityToProcessMap = {};
+        for (const dto of staticProcDefs) {
+          dto.calledFromActivityIds.forEach(
+            callerId => (callActivityToProcessMap[callerId] = dto)
+          );
+        }
+        const dynamicCallActivityToProcessesMap = {};
+        for (const dto of dynamicProcDefs) {
+          dto.calledFromActivityIds.forEach(callerId => {
+            if (dynamicCallActivityToProcessesMap[callerId]) {
+              dynamicCallActivityToProcessesMap[callerId].push(dto);
+            } else {
+              dynamicCallActivityToProcessesMap[callerId] = [dto];
+            }
+          });
+        }
+
+        for (const activity of callActivityFlowNodes) {
+          let redirectToId = callActivityToProcessMap[activity]
+            ? callActivityToProcessMap[activity].id
+            : undefined;
+          let toolTipTitle = redirectToId ? resolvable : notResolvable;
+          let clickListener = redirectToCalledDefinition;
+          if (!redirectToId && viewContext === 'runtime') {
+            /* We only link currently running dynamic call activities in the runtime view,
+            because the history view has no calledProcessDefinitionTable we can redirect to if there are different
+            process definitions called by one dynamic call activity.
+             */
+            if (dynamicCallActivityToProcessesMap[activity]) {
+              if (dynamicCallActivityToProcessesMap[activity].length > 1) {
+                redirectToId = activity;
+                clickListener = showCalledDefinitionsInTable;
+                toolTipTitle = dynamicMultipleResolve;
+              } else {
+                redirectToId =
+                  dynamicCallActivityToProcessesMap[activity][0].id;
+                toolTipTitle = dynamicResolve;
+              }
+            }
+          }
+
+          addOverlayForSingleElement(
+            overlaysNodes,
+            activity,
+            redirectToId,
+            overlays,
+            clickListener,
+            toolTipTitle,
+            $scope,
+            $timeout
+          );
+        }
+      };
```

Comment:
Even though it works technically and it optimizes the lines of code, I find it not so easy to read and understand.
I would write it as follows:
```suggestion
        for (const activityId of callActivityFlowNodes) {
          const callActivityProcess = callActivityToProcessMap[activityId];
          const dynamicCallActivityProcess =
            dynamicCallActivityToProcessesMap[activityId];
          const overlayProps = {
            overlaysNodes: {},
            activityId,
            overlays: control.getViewer().get('overlays'),
            $scope,
            $timeout
          };
          if (
            !callActivityProcess &&
            dynamicCallActivityProcess &&
            viewContext === 'runtime'
          ) {
            if (dynamicCallActivityProcess.length > 1) {
              addOverlayForSingleElement({
                redirectionTarget: activityId,
                clickListener: showCalledDefinitionsInTable,
                tooltipTitle: dynamicMultipleResolveTooltip,
                ...overlayProps
              });
            } else {
              addOverlayForSingleElement({
                redirectionTarget: dynamicCallActivityProcess[0].id,
                clickListener: redirectToCalledDefinition,
                tooltipTitle: dynamicResolveTooltip,
                ...overlayProps
              });
            }
          } else if (callActivityProcess) {
            addOverlayForSingleElement({
              redirectionTarget: callActivityProcess.id,
              clickListener: redirectToCalledDefinition,
              tooltipTitle: resolvableTooltip,
              ...overlayProps
            });
          } else {
            addOverlayForSingleElement({
              tooltipTitle: notResolvableTooltip,
              ...overlayProps
            });
          }
        }
```

Reply:
Yes, your approach looks easier to understand. Should I add the comment regarding why we restrict it to runtime back in or is that superficial?

Reply:
Sure. I think comments make the code even more readable.

Reply:
One more thing I accidentally introduced. Let's move:
```javascript
        const overlayProps = {
          overlaysNodes: {},
          overlays: control.getViewer().get('overlays'),
          $scope,
          $timeout
        };
```
Out of the for loop and add the `activityId` to each `addOverlayForSingleElement` method call. Otherwise, the same overlays might be added multiple times, right?

Reply:
Also, constructing this object once is cheaper.

Reply:
Don't forget to adjust the function signature in `callActivityOverlay.js`:
```javascript
function addOverlayForSingleElement({
  overlaysNodes,
  activityId,
  redirectionTarget,
  overlays,
  clickListener,
  tooltipTitle,
  $scope,
  $timeout
}) {
```
Is this function called somewhere else as well?